### PR TITLE
chore(migrate): Use the migration engine JSON-RPC printing API

### PR DESCRIPTION
**Pull request by tomhoule**
https://github.com/prisma/prisma-engines/pull/2566
todo: explain better

Note:
https://www.jsonrpc.org/specification

### CODE REVIEW

File: packages/migrate/src/MigrateEngine.ts
Comment (tomhoule) [inline packages/migrate/src/MigrateEngine.ts:142]: Deleted because I'm confident this code path is never taken, based on what the ME returns.

**Comment by Jolg42** [comment]
@tomhoule Let's sync about this tomorrow?

**Comment by tomhoule** [comment]
yep, good idea, let's do that :+1:

File: packages/migrate/src/MigrateEngine.ts
Comment (Jolg42) [inline packages/migrate/src/MigrateEngine.ts:141]:
```suggestion
      // It's a response
      if (result.id) {
```

File: packages/migrate/src/MigrateEngine.ts
Comment (Jolg42) [inline packages/migrate/src/MigrateEngine.ts:151]:
```suggestion
          if (result.method === "print" && result.params?.content) {
```

File: packages/migrate/src/MigrateEngine.ts
Comment (Jolg42) [inline packages/migrate/src/MigrateEngine.ts:152]:
```suggestion
              process.stdout.write(result.params.content)
```

File: packages/migrate/src/__tests__/MigrateDev.test.ts
Comment (Jolg42) [inline packages/migrate/src/__tests__/MigrateDev.test.ts:439]: @tomhoule this doesn't look good

File: packages/migrate/src/commands/MigrateDiff.ts
Comment (Jolg42) [inline packages/migrate/src/commands/MigrateDiff.ts:252]: @tomhoule we really need to clean this up soon and pass the diff in the response 🙈

File: packages/migrate/src/__tests__/MigrateDev.test.ts
Comment (tomhoule) [inline packages/migrate/src/__tests__/MigrateDev.test.ts:632]: Is this expected? I'm a bit puzzled

File: packages/migrate/src/commands/MigrateDiff.ts
Comment (tomhoule) [reply packages/migrate/src/commands/MigrateDiff.ts:252]: I am pretty sure asynchronously waiting here is not necessary. What we did before is send the notifications and the response on different channels (stderr and stdout), but with the companion PR that is already merged, the print notification is guaranteed to come before the response on stdout, so if the ordering of operations in the CLI matches the ordering of the JSON-RPC messages on stdout, there is no need for extra logic.

If we want to impose more synchronization and delays (e.g. if the log call and the response are reordered by the event loop), we could make the notification a request — the CLI would then send a response (an ACK), and the engine would wait for that before sending its own response. Long term, I think we want to use this mechanism a lot more, not less, because it avoids spreading control flow across both sides of the RPC boundary.

File: packages/migrate/src/commands/MigrateDeploy.ts
Comment (tomhoule) [inline packages/migrate/src/commands/MigrateDeploy.ts:126]: (see other comment, but I think it's the same here)

File: packages/migrate/src/commands/MigrateDev.ts
Comment (tomhoule) [inline packages/migrate/src/commands/MigrateDev.ts:179]: same here

**Review by tomhoule — COMMENTED** [review]
Since these waits are not nice at all, let's find a solid solution instead. Can you confirm they are really needed with an up-to-date version of engines?

File: packages/migrate/src/__tests__/MigrateDev.test.ts
Comment (Jolg42) [reply packages/migrate/src/__tests__/MigrateDev.test.ts:632]: Jest snapshots do that for a reason we don't completely understand.
You can ignore

File: packages/migrate/src/commands/MigrateDiff.ts
Comment (millsp) [reply packages/migrate/src/commands/MigrateDiff.ts:252]: Thanks, that makes a lot of sense. Hopefully, this will resolve most of the flakyness we see today :100:

Comment (Jolg42) [reply packages/migrate/src/commands/MigrateDiff.ts]: @tomhoule the tests fail when we do not wait, see
https://github.com/prisma/prisma/runs/5182960010?check_suite_focus=true

**Review by Jolg42 — APPROVED** [review]
Thanks for the comments,
Indeed, we can improve the communication of CLI - engine here.

I will merge this now to unblock the pipeline.

**Pull request merged**
